src/transmem.cpp

```diff
@@ -4,7 +4,333 @@
+        ptr2Link = &links.back();
+        ptr2SrcFrame->addLink(ptr2Link);
+    }
+
+    // add the transformation to the link
+    if(!ptr2Link->addTransformation(srcFrame, StampedTransformation{tstamp, qrot, qtrans})){
+        // TODO debug message: entry not stored since entry is to old
+    }
+
+    // release the lock
+    lock.unlock();
+
+    return;
+}
+
+void TransMem::registerLink(const FrameID &srcFrame, const FrameID &destFrame, const Timestamp &tstamp, const QMatrix4x4 &trans) {
+
+   float data[]{trans(0,0),trans(0,1),trans(0,2),
+                trans(1,0),trans(1,1),trans(1,2),
+                trans(2,0),trans(2,1),trans(2,2)};
+
+   QMatrix3x3 rM(data);
+
+   // TODO: check if the rotation matrix is normal, emit a warning if not
+
+   registerLink(srcFrame, destFrame, tstamp, QQuaternion::fromRotationMatrix(rM), QQuaternion(0, trans(0,3), trans(1,3), trans(2,3)));
+
+}
+
+void TransMem::writeJSON(QJsonObject &json) const {
+
+    QJsonArray frameObjects;
+    for(Frame f: frames){
+        QJsonObject frameObject;
+        f.writeJSON(frameObject);
+        frameObjects.append(frameObject);
+    }
+
+    QJsonArray linkObjects;
+    for(Link l: links){
+        QJsonObject linkObject;
+        l.writeJSON(linkObject);
+        linkObjects.append(linkObject);
+    }
+
+    json.insert("frames", frameObjects);
+    json.insert("links", linkObjects);
+
+}
+
+bool TransMem::shortestPath(Path &path){
+
+    Diijkstra diijkstra(frameID2Frame);
+    return diijkstra.calculateShortestPath(path);
+
+}
+
+void TransMem::dumpAsJSON() {
+
+    // TODO: add date and time to filename of dump
+
+    QJsonObject transmemObject;
+    lock.lock(); writeJSON(transmemObject); lock.unlock();
+
+    dumpJSONfile("TransMemDump", transmemObject);
+
+    return;
+}
+
+void TransMem::dumpJSONfile(const QString &path, const QJsonObject &json) const {
+
+    QFile file(path+".json");
+    if(!file.open(QIODevice::WriteOnly)){
+        // TODO: error handling
+        return;
+    }
+
+    QJsonDocument saveJSON(json);
+    file.write(saveJSON.toJson());
+
+    file.close();
+    if(file.error()){
+        // TODO: error handling
+        return;
+    }
+
+}
+
+void TransMem::dumpPathAsJSON(const Path &p){
+
+    // TODO:: add date and time to filname of dump
+
+   QJsonObject pathObject;
+   lock.lock(); p.writeJSON(pathObject); lock.unlock();
+
+   dumpJSONfile("PathDump", pathObject);
+
+   return;
+}
+
+void TransMem::dumpAsGraphML() {
+
+   GMLWriter writer;
+
+   lock.lock(); writer.write(this); lock.unlock();
+
+}
+
+QMatrix4x4 TransMem::getLink(const FrameID &srcFrame, const FrameID &destFrame, const Timestamp &tstamp) {
+
+    if(srcFrame == destFrame)
+        throw std::invalid_argument("Not allowed to query for link if source frame is equal to destination frame.");
+
+    // get the lock
+    lock.lock();
+
+    // search for shortest path between source frame and  destination frame
+    Path p{srcFrame, destFrame, std::vector<Link*>()};
```

Comment:
You are right. I initially added it because if omitted I get a "missing initializer for member" warning. My goal was to stay warning free throughout the whole development. @stephanemagnenat what do you think, how important is it to stay "warning free"?